opt: panic: lookup for ComparisonExpr ((ARRAY[('1970-01-10 03:46:29.00008+00:00')[string],('1970-01-03 13:05:19.000228+00:00')[string]])[string[]] IS NOT DISTINCT FROM (ARRAY[])[varbit[]])[bool]'s CmpOp failed

Entering the following transcript causes a crash:

```
CREATE TABLE table1 (col0 INT8[], col1 INET NOT NULL, col2 FLOAT8 NOT NULL, col3 TIMESTAMPTZ, col4 INT8[] NOT NULL, col5 NAME NOT NULL, col6 UUID, col7 INT8 NULL, col8 TIMESTAMP NULL, col9 INT8[], col10 INT8[], col11 UUID NULL, col12 STRING, col13 INET, col14 DECIMAL, col15 BYTES NULL, col16 INT8 NOT NULL, col17 FLOAT8 NULL, col18 NAME, FAMILY fam0 (col2, col15), FAMILY fam1 (col10, col4, col1), FAMILY fam2 (col17, col0), FAMILY fam3 (col13, col18, col8, col3, col16), FAMILY fam4 (col12, col6, col14, col5, col11, col7), FAMILY fam5 (col9), PRIMARY KEY (col15 DESC), INDEX (col13, col16 DESC, col8 DESC, col17, col6 DESC, col1 ASC, col12 DESC, col11 ASC, col7 ASC, col5 DESC, col14, col15, col2 ASC, col3 DESC, col18), UNIQUE (col11 ASC, col5 ASC, col17 DESC, col15 ASC, col7, col14, col6, col12 ASC, col18 ASC, col2 DESC, col8 ASC, col13), UNIQUE (col16, col18 DESC, col3, col8 DESC, col13, col7 ASC, col17, col2 ASC, col1, col12 ASC), INDEX (col18 ASC, col1 ASC, col3, col16 DESC, col8 DESC, col12 ASC, col6 DESC, col2 ASC, col15 ASC, col11 ASC));
CREATE TABLE table4 (col0 INT8 NOT NULL, col1 INT8[], FAMILY fam0 (col0), FAMILY fam1 (col1), PRIMARY KEY (col0 ASC), INDEX (col0 DESC), INDEX (col0 ASC), INDEX (col0 ASC), INDEX (col0 DESC), INDEX (col0 DESC), UNIQUE (col0 DESC), UNIQUE (col0 ASC));
SELECT CASE WHEN (SELECT false FROM table4 AS tab_2 WHERE (ARRAY['1970-01-10 03:46:29.00008+00:00', '1970-01-03 13:05:19.000228+00:00'] IS NOT DISTINCT FROM ARRAY[]) LIMIT 1) THEN NULL ELSE B'010101000110000100001100110100111010011000100001000101100000100111011101001001011110000010100100' END FROM table1 AS tab_1 WHERE (ARRAY[B'001010', B'00000110001000110010100111110100110010110'] != ARRAY[]) LIMIT 10;
```

```
panic: lookup for ComparisonExpr ((ARRAY[('1970-01-10 03:46:29.00008+00:00')[string],('1970-01-03 13:05:19.000228+00:00')[string]])[string[]] IS NOT DISTINCT FROM (ARRAY[])[varbit[]])[bool]'s CmpOp failed [recovered]
	panic: panic while executing 1 statements: SELECT CASE WHEN (SELECT _ FROM _ AS _ WHERE (ARRAY[_, _] IS NOT DISTINCT FROM ARRAY[]) LIMIT _) THEN _ ELSE _ END FROM _ AS _ WHERE (ARRAY[_, _] != ARRAY[]) LIMIT _; caused by lookup for ComparisonExpr ((ARRAY[('1970-01-10 03:46:29.00008+00:00')[string],('1970-01-03 13:05:19.000228+00:00')[string]])[string[]] IS NOT DISTINCT FROM (ARRAY[])[varbit[]])[bool]'s CmpOp failed

goroutine 389 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc000396000, 0x73f01e0, 0xc00094c240, 0x675e9e0, 0xc0004f96d0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:711 +0x36d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc000396000, 0x73f01e0, 0xc00094c240)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:423 +0x61
panic(0x675e9e0, 0xc0004f96d0)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ComparisonExpr).memoizeFn(0xc00188d9f0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/expr.go:516 +0x4a2
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.NewTypedComparisonExpr(0x13, 0x740d160, 0xc00189fa40, 0x740d160, 0xc0018128d0, 0xc0018128d0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/expr.go:430 +0xf7
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildComparison(0xc00188d9a0, 0xc001077640, 0x740b420, 0xc000369be0, 0x6a0e5c0, 0x740b420, 0x6a0e5c0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:228 +0x1ff
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc00188d9a0, 0xc001077640, 0x740b420, 0xc000369be0, 0xc000369be0, 0x91f8620, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:94 +0xa6
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildBoolean(0xc00188d9a0, 0xc001077640, 0x740b060, 0xc001407080, 0x6a487c0, 0x740b060, 0xc000146c80, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:209 +0x789
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc00188d9a0, 0xc001077640, 0x740b060, 0xc001407080, 0xc001407080, 0xc001077640, 0xc000533280, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:94 +0xa6
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildBoolean(0xc00188d9a0, 0xc001077640, 0x740b000, 0xc0014d48f8, 0x0, 0x0, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:178 +0xdb
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc00188d9a0, 0xc001077640, 0x740b000, 0xc0014d48f8, 0x0, 0x0, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:94 +0xa6
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildSelect(0xc00188d9a0, 0xc0014d48e8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x91f8620, 0x91f8620)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:400 +0x194
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc00188d9a0, 0x74281e0, 0xc0014d48e8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x4ac2de4, 0x91f9060)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:164 +0xba7
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildLimitOffset(0xc00188d9a0, 0x7427a60, 0xc00081fb68, 0x0, 0x0, 0x0, 0x0, 0x0, 0x73b8760, 0x400b9ab)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:856 +0xd9
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc00188d9a0, 0x7427a60, 0xc00081fb68, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:176 +0x858
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildProject(0xc00188d9a0, 0xc00081fd68, 0x0, 0x0, 0x0, 0x0, 0x0, 0x4046017, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:437 +0xa5
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc00188d9a0, 0x7427d80, 0xc00081fd68, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00003a000, 0x6930d00)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:167 +0x792
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).build(0xc00188d9a0, 0x73fe6a0, 0xc00081fd68, 0x73fe6a0, 0xc00081fd68, 0x740d220, 0x73cad20)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:62 +0xab
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildSubquery(0xc00188d9a0, 0xc0010775c0, 0x740c260, 0xc00188d8b0, 0xc600d80, 0x0, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:544 +0x12d
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc00188d9a0, 0xc0010775c0, 0x740c260, 0xc00188d8b0, 0x1, 0x1, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:94 +0xa6
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildCase(0xc00188d9a0, 0xc0010775c0, 0x740a820, 0xc00188d900, 0x0, 0x0, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:295 +0x1ca
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc00188d9a0, 0xc0010775c0, 0x740a820, 0xc00188d900, 0x0, 0x0, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:94 +0xa6
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildProject(0xc00188d9a0, 0xc0017a0568, 0x0, 0x0, 0x0, 0x0, 0x0, 0x91a9510, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:453 +0x2de
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc00188d9a0, 0x7427d80, 0xc0017a0568, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0016e2990, 0x400ec6f)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:167 +0x792
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).build(0xc00188d9a0, 0x73fe6a0, 0xc0017a0568, 0x0, 0xc0016e29c0, 0x400f4f8, 0x50)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:62 +0xab
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).Build(0xc00188d9a0, 0xc00188d9a0, 0xc0016ec1b0, 0xc001842f20, 0xc000513c00)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:50 +0x41
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc000396450, 0x73f02a0, 0xc0016ec1b0, 0x73f5d20, 0xc0013dfb80, 0xc000513cdb, 0x195, 0x0, 0x0, 0x0, ...)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:140 +0x1e4
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).makeExecPlan(0xc000396000, 0x73f02a0, 0xc0016ec1b0, 0x73f5d20, 0xc0013dfb80, 0xc000513cdb, 0x195, 0x0, 0x0, 0x0, ...)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:888 +0x162
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc000396000, 0x73f02a0, 0xc0016ec1b0, 0x73f5d20, 0xc0013dfb80, 0xc000513cdb, 0x195, 0x0, 0x0, 0x0, ...)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:800 +0x170
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc000396000, 0x73f02a0, 0xc0016ec1b0, 0x73f5d20, 0xc0013dfb80, 0xc000513cdb, 0x195, 0x0, 0x0, 0x0, ...)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:437 +0xe85
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc000396000, 0x73f02a0, 0xc0016ec1b0, 0x73f5d20, 0xc0013dfb80, 0xc000513cdb, 0x195, 0x0, 0x0, 0x0, ...)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:93 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc000396000, 0x73f01e0, 0xc00094c240, 0xc0000f6c38, 0x5400, 0x15000, 0xc0000f6cd0, 0xc000428180, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1172 +0x21e5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000171ba0, 0x73f01e0, 0xc00094c240, 0xc000396000, 0x5400, 0x15000, 0xc0000f6cd0, 0xc000428180, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:425 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc000171ba0, 0x73f01e0, 0xc00094c240, 0xc000396000, 0x5400, 0x15000, 0xc0000f6cd0, 0xc000428180, 0xc000428190, 0xc000340134)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:319 +0x81
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:318 +0x1033
```

--- a/pkg/sql/opt/memo/interner.go
+++ b/pkg/sql/opt/memo/interner.go
@@ -363,16 +363,32 @@ func (h *hasher) HashDatum(val tree.Datum) {
 	case *tree.DJSON:
 		h.HashString(t.String())
 	case *tree.DTuple:
-		for _, d := range t.D {
-			h.HashDatum(d)
-		}
-		h.HashDatumType(t.ResolvedType())
+		// If labels are present, then hash of tuple's static type is needed to
+		// disambiguate when everything is the same except labels.
+		alwaysHashType := len(t.ResolvedType().(types.TTuple).Labels) != 0
+		h.hashDatumsWithType(t.D, t.ResolvedType(), alwaysHashType)
 	case *tree.DArray:
-		for _, d := range t.Array {
-			h.HashDatum(d)
-		}
+		// If the array is empty, then hash of tuple's static type is needed to
+		// disambiguate.
+		alwaysHashType := len(t.Array) == 0
+		h.hashDatumsWithType(t.Array, t.ResolvedType(), alwaysHashType)
 	default:
-		h.HashBytes(encodeDatum(h.bytes[:0], val))
+		h.bytes = encodeDatum(h.bytes[:0], val)
+		h.HashBytes(h.bytes)
+	}
+}
+
+func (h *hasher) hashDatumsWithType(datums tree.Datums, typ types.T, alwaysHashType bool) {
+	for _, d := range datums {
+		if d == tree.DNull {
+			// At least one NULL exists, so need to compare static types (e.g. a
+			// NULL::int is indistinguishable from NULL::string).
+			alwaysHashType = true
+		}
+		h.HashDatum(d)
+	}
+	if alwaysHashType {
+		h.HashDatumType(typ)
 	}
 }
 
@@ -381,9 +397,10 @@ func (h *hasher) HashDatumType(val types.T) {
 }
 
 func (h *hasher) HashColType(val coltypes.T) {
-	buf := bytes.NewBuffer(h.bytes)
+	buf := bytes.NewBuffer(h.bytes[:0])
 	val.Format(buf, lex.EncNoFlags)
-	h.HashBytes(buf.Bytes())
+	h.bytes = buf.Bytes()
+	h.HashBytes(h.bytes)
 }
 
 func (h *hasher) HashTypedExpr(val tree.TypedExpr) {
@@ -579,11 +596,13 @@ func (h *hasher) IsDatumTypeEqual(l, r types.T) bool {
 }
 
 func (h *hasher) IsColTypeEqual(l, r coltypes.T) bool {
-	lbuf := bytes.NewBuffer(h.bytes)
+	lbuf := bytes.NewBuffer(h.bytes[:0])
 	l.Format(lbuf, lex.EncNoFlags)
-	rbuf := bytes.NewBuffer(h.bytes2)
+	rbuf := bytes.NewBuffer(h.bytes2[:0])
 	r.Format(rbuf, lex.EncNoFlags)
-	return bytes.Equal(lbuf.Bytes(), rbuf.Bytes())
+	h.bytes = lbuf.Bytes()
+	h.bytes2 = rbuf.Bytes()
+	return bytes.Equal(h.bytes, h.bytes2)
 }
 
 func (h *hasher) IsDatumEqual(l, r tree.Datum) bool {
@@ -622,35 +641,54 @@ func (h *hasher) IsDatumEqual(l, r tree.Datum) bool {
 		}
 	case *tree.DTuple:
 		if rt, ok := r.(*tree.DTuple); ok {
-			if len(lt.D) != len(rt.D) {
+			// Compare datums and then compare static types if nulls or labels
+			// are present.
+			ltyp := lt.ResolvedType().(types.TTuple)
+			rtyp := rt.ResolvedType().(types.TTuple)
+			if !h.areDatumsWithTypeEqual(lt.D, rt.D, ltyp, rtyp) {
 				return false
 			}
-			for i := range lt.D {
-				if !h.IsDatumEqual(lt.D[i], rt.D[i]) {
-					return false
-				}
-			}
-			return h.IsDatumTypeEqual(l.ResolvedType(), r.ResolvedType())
+			return len(ltyp.Labels) == 0 || h.IsDatumTypeEqual(ltyp, rtyp)
 		}
 	case *tree.DArray:
 		if rt, ok := r.(*tree.DArray); ok {
-			if len(lt.Array) != len(rt.Array) {
+			// Compare datums and then compare static types if nulls are present
+			// or if arrays are empty.
+			ltyp := lt.ResolvedType()
+			rtyp := rt.ResolvedType()
+			if !h.areDatumsWithTypeEqual(lt.Array, rt.Array, ltyp, rtyp) {
 				return false
 			}
-			for i := range lt.Array {
-				if !h.IsDatumEqual(lt.Array[i], rt.Array[i]) {
-					return false
-				}
-			}
-			return true
+			return len(lt.Array) != 0 || h.IsDatumTypeEqual(ltyp, rtyp)
 		}
 	default:
-		lb := encodeDatum(h.bytes[:0], l)
-		rb := encodeDatum(h.bytes2[:0], r)
-		return bytes.Equal(lb, rb)
+		h.bytes = encodeDatum(h.bytes[:0], l)
+		h.bytes2 = encodeDatum(h.bytes2[:0], r)
+		return bytes.Equal(h.bytes, h.bytes2)
 	}
 
 	return false
+}
+
+func (h *hasher) areDatumsWithTypeEqual(ldatums, rdatums tree.Datums, ltyp, rtyp types.T) bool {
+	if len(ldatums) != len(rdatums) {
+		return false
+	}
+	foundNull := false
+	for i := range ldatums {
+		if !h.IsDatumEqual(ldatums[i], rdatums[i]) {
+			return false
+		}
+		if ldatums[i] == tree.DNull {
+			// At least one NULL exists, so need to compare static types (e.g. a
+			// NULL::int is indistinguishable from NULL::string).
+			foundNull = true
+		}
+	}
+	if foundNull {
+		return h.IsDatumTypeEqual(ltyp, rtyp)
+	}
+	return true
 }
 
 func (h *hasher) IsTypedExprEqual(l, r tree.TypedExpr) bool {

--- a/pkg/sql/opt/memo/interner_test.go
+++ b/pkg/sql/opt/memo/interner_test.go
@@ -42,11 +42,16 @@ func TestInterner(t *testing.T) {
 	tupTyp2 := types.TTuple{Types: []types.T{types.Int, types.String}, Labels: []string{"a", "b"}}
 	tupTyp3 := types.TTuple{Types: []types.T{types.Int, types.String}}
 	tupTyp4 := types.TTuple{Types: []types.T{types.Int, types.String, types.Bool}}
+	tupTyp5 := types.TTuple{Types: []types.T{types.Int, types.String}, Labels: []string{"c", "d"}}
+	tupTyp6 := types.TTuple{Types: []types.T{types.String, types.Int}, Labels: []string{"c", "d"}}
 
 	tup1 := tree.NewDTuple(tupTyp1, tree.NewDInt(100), tree.NewDString("foo"))
 	tup2 := tree.NewDTuple(tupTyp2, tree.NewDInt(100), tree.NewDString("foo"))
 	tup3 := tree.NewDTuple(tupTyp3, tree.NewDInt(100), tree.NewDString("foo"))
 	tup4 := tree.NewDTuple(tupTyp4, tree.NewDInt(100), tree.NewDString("foo"), tree.DBoolTrue)
+	tup5 := tree.NewDTuple(tupTyp5, tree.NewDInt(100), tree.NewDString("foo"))
+	tup6 := tree.NewDTuple(tupTyp5, tree.DNull, tree.DNull)
+	tup7 := tree.NewDTuple(tupTyp6, tree.DNull, tree.DNull)
 
 	arr1 := tree.NewDArray(tupTyp1)
 	arr1.Array = tree.Datums{tup1, tup2}
@@ -54,6 +59,14 @@ func TestInterner(t *testing.T) {
 	arr2.Array = tree.Datums{tup2, tup1}
 	arr3 := tree.NewDArray(tupTyp3)
 	arr3.Array = tree.Datums{tup2, tup3}
+	arr4 := tree.NewDArray(types.Int)
+	arr4.Array = tree.Datums{tree.DNull}
+	arr5 := tree.NewDArray(types.String)
+	arr5.Array = tree.Datums{tree.DNull}
+	arr6 := tree.NewDArray(types.Int)
+	arr6.Array = tree.Datums{}
+	arr7 := tree.NewDArray(types.String)
+	arr7.Array = tree.Datums{}
 
 	dec1, _ := tree.ParseDDecimal("1.0")
 	dec2, _ := tree.ParseDDecimal("1.0")
@@ -187,9 +200,14 @@ func TestInterner(t *testing.T) {
 			{val1: tup1, val2: tup2, equal: true},
 			{val1: tup2, val2: tup3, equal: false},
 			{val1: tup3, val2: tup4, equal: false},
+			{val1: tup1, val2: tup5, equal: false},
+			{val1: tup6, val2: tup7, equal: false},
 
 			{val1: arr1, val2: arr2, equal: true},
 			{val1: arr2, val2: arr3, equal: false},
+			{val1: arr4, val2: arr5, equal: false},
+			{val1: arr4, val2: arr6, equal: false},
+			{val1: arr6, val2: arr7, equal: false},
 
 			{val1: dec1, val2: dec2, equal: true},
 			{val1: dec2, val2: dec3, equal: false},

